### PR TITLE
Add smoke test for JWT realm wiring

### PR DESCRIPTION
**Pull request by justincr-elastic**
Verify JWT realm settings can be loaded by ES server, and queried by ES client.

Verify REST client can submit JWT and client secret, and ES server JWT realm will perform successful authentication.

### CODE REVIEW

**Comment by justincr-elastic** [comment]
Hi @ywangd, I think coding is done, but I am running into two issues.

1. JWT string value is much longer than 140 characters. Automatic formatting does not fix it.
`./gradlew       spotlessApply -Des.format.everything checkstyle`

2. I generated JWT realm settings, JWT, and client secret in JwtRealmTests.testCreateJwtSmokeTestRealm(). That test verifies authentication succeeded too. However, when I put those JWT realm settings, JWT, and client secret, in the smoke test it fails to authenticate. I am not getting any server logs, only client logs, so I am unclear why it fails or how to diagnose the issue.

3. I left some debug logs in the code with TODO to remove them. I will after I figure out what is broken in 2.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ */
+public class JwtRealmAuthIT extends SecurityRealmSmokeTestCase {
+    private static final Logger LOGGER = LogManager.getLogger(JwtRealmAuthIT.class);
+
+    // Declared in build.gradle
+    private static final String USERNAME = "security_test_user";
+    private static final String ROLE_NAME = "security_test_role";
+    private static final String HEADER_CLIENT_SECRET = "client-shared-secret-string";
+    private static final String HEADER_JWT =
+        "eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZXMiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE";
```
Comment (justincr-elastic) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:31]: Spotless doesn't know how to fix this. I tried putting it in a file and loading it, but I get read access denied.

Is there another way to keep it intact, without breaking it up just to satisfy checkstyle?

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        // TODO Remove debug log
+        LOGGER.info(
+            JwtRealm.HEADER_CLIENT_AUTHENTICATION
+                + ": "
+                + JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_SHARED_SECRET
+                + " "
+                + HEADER_CLIENT_SECRET
+        );
+        LOGGER.info(JwtRealm.HEADER_END_USER_AUTHENTICATION + ": " + JwtRealm.HEADER_END_USER_AUTHENTICATION_SCHEME + " " + HEADER_JWT);
+        LOGGER.info(options.build().toString());
```
Comment (justincr-elastic) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:50]: I expected the test to succeeded. I will remove these logs after we figure out why the test is failing.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java [hunk 3]
```diff
@@ -107,16 +107,18 @@ public JwtRealm(final RealmConfig realmConfig, final SSLService sslService, fina
             this.httpClient = null; // no setting means no HTTP client
         }
 
-        this.jwksAlgsHmac = this.parseJwksAlgsHmac(); // not reloadable
-        this.jwksAlgsPkc = this.parseJwksAlgsPkc(false); // reloadable
+        this.parseJwksAlgsHmac();
+        this.parseJwksAlgsPkc();
+        this.verifyAnyAvailableJwkAndAlgPair();
```
Comment (justincr-elastic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java:112]: Logging bug fix. If HMAC key filtered out (ex: too small), diagnostic log in PKC JWKSet method was being skipped because it was empty. I moved the checks to a separate method to ensure it gets printed. I found this issue while trying to generate the static realm config for the smoke test.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java [hunk 4]
```diff
@@ -108,6 +104,122 @@ public void shutdown() throws Exception {
...
+    /**
+     * Generate a static JWT realm config, and request headers, which the realm can successfully authenticate.
+     *
+     * Print these artifacts, so they can be imported into this integration test project for verifying loading of JWT realm config:
+     *  - x-pack/plugin/security/qa/smoke-test-all-realms/build.gradle
+     *
+     * If something needs tweaking, change the code here and re-run it. The new artifacts can be used to update that smoke test project.
+     * @throws Exception Error if something goes wrong.
+     */
+    public void testCreateJwtSmokeTestRealm() throws Exception {
```
Comment (justincr-elastic) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java:116]: The output of this test is the static JWT realm config, client secret, and signed JWT, all to be used in the realm config smoke test. The output of this test is saved in the README.asciidoc. The result of this test is deterministic, re-running it produces the same realm config, client secret, and signed JWT.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java [hunk 5]
```diff
@@ -328,15 +378,21 @@ public static OctetSequenceKey randomJwkHmacOidc(final JWSAlgorithm jwsAlgorithm
...
-        final OctetSequenceKey.Builder utf8HmacKeyBuilder = new OctetSequenceKey.Builder(base64AsUtf8);
-        utf8HmacKeyBuilder.keyID(hmacKey.getKeyID()); // Copy null attribute is OK (no-op)
-        utf8HmacKeyBuilder.algorithm(hmacKey.getAlgorithm());
-        utf8HmacKeyBuilder.keyUse(hmacKey.getKeyUse());
-        utf8HmacKeyBuilder.keyOperations(hmacKey.getKeyOperations());
-        utf8HmacKeyBuilder.keyStore(hmacKey.getKeyStore());
-        return utf8HmacKeyBuilder.build();
+        final String passwordKey;
+        if (randomBoolean()) {
+            final Base64URL hmacKeyBytesBase64 = hmacKey.getKeyValue(); // Random bytes => 8 bits/byte search space
```
Comment (justincr-elastic) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java:383]: I previously used Base64 of random bytes as a new plain text string. It worked, but gave misleading impressing base64 was required. It is not, so I added some fuzzing to ensure random letter passwords can be used too.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java [hunk 6]
```diff
@@ -363,48 +419,28 @@ public static OctetSequenceKey jwkHmacRemoveAttributes(final OctetSequenceKey hm
...
-        return JwtValidateUtil.signJwt(jwsSigner, unsignedJwt);
-    }
-
-    public SecureString buildJWT(final JWSHeader header, final JWTClaimsSet claims, final Base64URL signature) throws ParseException {
-        final SignedJWT signedJwt = new SignedJWT(header.toBase64URL(), claims.toPayload().toBase64URL(), signature);
-        return new SecureString(signedJwt.serialize().toCharArray());
-    }
-
-    public static SignedJWT randomValidJwsHeaderAndJwtClaimsSet(
+    public static SignedJWT buildUnsignedJwt(
```
Comment (justincr-elastic) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtTestCase.java:422]: Refactored to pass in all values, instead of generating some as random. This was required to generate a reproducible JWT for the realm config smoke test, with fixed claims like exp claim time 2099/1/1 12:00:00.000.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java (hunk 1 above)
Comment (ywangd) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:31]: We can break this long string into a concatenation, e.g. 
```
        final String HEADER_JWT = "eyJhbGciOiJIUzI1NiJ9."
            + "eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZX"
            + "MiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA"
            + "5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE";
```

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/README.asciidoc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+PKC JWKSet contents
+===
+Not found.
+
+===
+elasticsearch-keystore secure settings
+===
+xpack.security.authc.realms.jwt.jwt8.hmac_key: hmac-oidc-key-string-for-hs256-algorithm
+xpack.security.authc.realms.jwt.jwt8.client_authentication.shared_secret: client-shared-secret-string
+
```
Comment (ywangd) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/README.asciidoc:48]: I think we only need this section in this file. Rest is noise and it's easier to understand the point without it.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/clientsecret.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+client-shared-secret-string
```
Comment (ywangd) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/clientsecret.txt:1]: I don't think this file is needed. right?

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/jwt.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiJhdWQ4Iiwic3ViIjoic2VjdXJpdHlfdGVzdF91c2VyIiwicm9sZXMiOiJbc2VjdXJpdHlfdGVzdF9yb2xlXSIsImlzcyI6ImlzczgiLCJleHAiOjQwNzA5MDg4MDAsImlhdCI6OTQ2Njg0ODAwfQ.YbMbSEY8j3BdE_M71np-5Q9DFHGcjZcu7D4Kk1Ji0wE
```
Comment (ywangd) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/jwt.txt:1]: Same here. This file is not needed?

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java [hunk 10]
```diff
@@ -717,7 +717,7 @@ private void multipleRealmsAuthenticateJwtHelper(
                 assertThat(new TreeSet<>(Arrays.asList(user.roles())), equalTo(new TreeSet<>(Arrays.asList(authenticatedUser.roles()))));
                 if (jwtIssuerAndRealm.realm.delegatedAuthorizationSupport.hasDelegation()) {
                     assertThat(user.metadata(), is(equalTo(authenticatedUser.metadata()))); // delegated authz returns user's
-                                                                                            // metadata
+                    // metadata
```
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java:720]: Err can we revert this change. The format looks off to me. Or make the trailing comment into its own line.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java [hunk 11]
```diff
@@ -409,6 +409,11 @@
         "oidc",
         License.OperationMode.PLATINUM
     );
+    public static final LicensedFeature.Persistent JWT_REALM_FEATURE = LicensedFeature.persistent(
+        REALMS_FEATURE_FAMILY,
+        "jwt",
+        License.OperationMode.PLATINUM
+    );
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:416]: I guess you haven't update the PR yet because this file needs the extra Rest Header change.

**Review by ywangd — COMMENTED** [review]
This PR looks promising. I have a few comments besides what we discussed in synchronously. Please request review again once you make it formal. Thanks!

File: .idea/checkstyle-idea.xml [hunk 12]
```diff
@@ -16,4 +16,4 @@
       </map>
     </option>
   </component>
-</project>
+</project>
```
[the remark this answers is not in the document]
Comment (justincr-elastic) [reply .idea/checkstyle-idea.xml]: Resolved.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java (hunk 1 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java]: Resolved. I did something similar.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/README.asciidoc (hunk 7 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/README.asciidoc:27]: Resolved.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/clientsecret.txt (hunk 8 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/clientsecret.txt]: Resolved. Removed file.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/jwt.txt (hunk 9 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/resources/jwt/jwt.txt]: Resolved. Removed file.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java (hunk 11 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:417]: Resolved. Yes, when we found the issue it close to midnight for me. I added the request header to the accept list first thing next morning in commit 0a50dc9e5d534ed1a4a02cce1d9e68e80472a20c.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java (hunk 10 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealmTests.java]: Resolved. I have no idea what that comment wrapped to the next line. Fixed it.

**Comment by justincr-elastic** [comment]
I am curious why are there two smoke tests for realm config?
- x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/RealmInfoIT.java
- x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/SecurityRealmSettingsTests.java

**Comment by tvernum** [comment]
I think we can consider `SecurityRealmSettingsTests` to be obsolete. The QA test is better because it actually starts a real node.
I think it's worth opening an issue to review whether the QA test covers everything that is in `SecurityRealmSettingsTests` and then delete `SecurityRealmSettingsTests`.

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            .addHeader(
+                JwtRealm.HEADER_CLIENT_AUTHENTICATION,
+                JwtRealmSettings.CLIENT_AUTHENTICATION_TYPE_SHARED_SECRET + " " + HEADER_CLIENT_SECRET
+            )
+            .addHeader(JwtRealm.HEADER_END_USER_AUTHENTICATION, JwtRealm.HEADER_END_USER_AUTHENTICATION_SCHEME + " " + HEADER_JWT);
+
+        final Map<String, Object> authenticate = super.authenticate(options);
+        assertUsername(authenticate, USERNAME);
+        assertRealm(authenticate, "jwt", "jwt8");
+        assertRoles(authenticate); // empty list
```
Comment (ywangd) [inline x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:45]: So we decided to go without role mapping? I am ok with it since this smoke test is more about the realm itself. Just to be sure.

**Review by ywangd — APPROVED** [review]
LGTM

File: x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java (hunk 13 above)
Comment (justincr-elastic) [reply x-pack/plugin/security/qa/smoke-test-all-realms/src/javaRestTest/java/org/elasticsearch/xpack/security/authc/JwtRealmAuthIT.java:45]: Correct, I went without role mapping. This PR was just about wiring in the config.

I considered spending more cycles during this PR to add a one-off authz check. However, Tim had recommended I create smaller, incremental PRs to keep the scope of each one tight. So, this PR is just for checking the config is wired in. The next planned PR is for integration testing, and that feels like a better place to verify role mapping for JWT realm.

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

Jenkins shutdown killed my test. Need to rerun it again.

**Comment by justincr-elastic** [comment]
@elasticmachine update branch

**Comment by justincr-elastic** [comment]
@elasticmachine run elasticsearch-ci/part-1
